### ISSUE TITLE
LSP running out of memory

### ISSUE THREAD

**Issue by varunrau** [yioneko/vtsls#193]
I'm observing that the LSP is frequently running out of memory on a large project. I have this in my `.vscode/settings.json` of the project so that VSCode's LSP works, but it doesn't seem like vtsls is observing that. Am I supposed to configure this setting differently?
```
  "typescript.tsserver.maxTsServerMemory": 13312,
```

Here's the error I see from `:LspLog`
```
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	"\n<--- Last few GCs --->\n\n[65593:0x7fcfc0040000]    38376 ms: Scavenge 4052.4 (4131.4) -> 4048.6 (4132.4) MB, 5.6 / 0.0 ms  (average mu = 0.851, current mu = 0.655) allocation failure; \n[65593:0x7fcfc0040000]    38393 ms: Scavenge 4053.9 (4132.8) -> 4050.7 (4135.6) MB, 6.6 / 0.0 ms  (average mu = 0.851, current mu = 0.655) allocation failure; \n[65593:0x7fcfc0040000]    39065 ms: Scavenge 4056.4 (4135.6) -> 4053.0 (4152.8) MB, 661.5 / 0.0 ms  (average mu = 0.851, current mu = 0.655) allocation failure; \n\n\n<--- JS stacktrace --->\n\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	"FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 1: 0x102612c45 node::Abort() (.cold.1) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 2: 0x10106d7b9 node::Abort() [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 3: 0x10106d98e node::OOMErrorHandler(char const*, bool) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 4: 0x1011faf43 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 5: 0x1013c3bf5 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 6: 0x1013c25d2 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 7: 0x1013b490a v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 8: 0x1013b5285 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 9: 0x1013976fe v8::internal::Factory::NewFillerObject(int, v8::internal::AllocationAlignment, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	"10: 0x1017cd69c v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	"11: 0x101bc2e79 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	"\n<--- Last few GCs --->\n\n[65585:0x7fbfe8078000]    38473 ms: Scavenge 4051.9 (4130.7) -> 4048.1 (4131.4) MB, 4.7 / 0.0 ms  (average mu = 0.843, current mu = 0.651) allocation failure; \n[65585:0x7fbfe8078000]    38490 ms: Scavenge 4053.4 (4131.8) -> 4050.3 (4135.1) MB, 6.2 / 0.0 ms  (average mu = 0.843, current mu = 0.651) allocation failure; \n[65585:0x7fbfe8078000]    39169 ms: Scavenge 4055.9 (4135.1) -> 4052.5 (4152.1) MB, 669.8 / 0.0 ms  (average mu = 0.843, current mu = 0.651) allocation failure; \n\n\n<--- JS stacktrace --->\n\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	"FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 1: 0x1040afc45 node::Abort() (.cold.1) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 2: 0x102b0a7b9 node::Abort() [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 3: 0x102b0a98e node::OOMErrorHandler(char const*, bool) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 4: 0x102c97f43 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 5: 0x102e60bf5 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 6: 0x102e5f5d2 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 7: 0x102e5190a v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 8: 0x102e52285 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	" 9: 0x102e346fe v8::internal::Factory::NewFillerObject(int, v8::internal::AllocationAlignment, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	"10: 0x10326a69c v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	"11: 0x10365fe79 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/usr/local/bin/node]\n"
[ERROR][2024-01-06 16:11:08] ...ools/process.lua:184	"process"	"tsserver"	"stderr"	"12: 0x11449f24c \n"
```

**Comment by yioneko** [comment]
Yes, the server doesn't support reading settings from `.vscode/` folder. The client is responsible for passing the configuration to the server and you should refer to the client's doc for that.

**Comment by yioneko** [comment]
I suspect the error messages you posted are from [typescript-tools.nvim](https://github.com/pmizio/typescript-tools.nvim/blob/f8c2e0b36b651c85f52ad5c5373ff8b07adc15a7/lua/typescript-tools/process.lua#L184) and not from here. They are in conflict with each other and it seems that you've reported the problem to the wrong place?

**Comment by varunrau** [comment]
Thanks for the confirmation! Interesting - I thought that it would be vtsls because that's the LSP client that is connected to this buffer.

I was looking in the vtsls docs for where to put the configuration, but I wasn't able to find it. Do you know where it's supposed to go?

I'm using neovim 0.10.0 for what it's worth.

```
Client: vtsls (id: 2, bufnr: [3])
 	filetypes:       javascript, javascriptreact, javascript.jsx, typescript, typescriptreact, typescript.tsx
 	autostart:       true
 	root directory:  /Users/vr/code/ts-playground
 	cmd:             /Users/vr/.local/share/nvim/mason/bin/vtsls --stdio
```

**Comment by yioneko** [comment]
It differs by client about how to pass configuration to server so there is no doc for it here. Specifically for neovim builtin LSP client, assuming that you are using nvim-lspconfig: `:h lspconfig-setup` and look for description about `settings` field.

TLDR example:
```lua
require("lspconfig").vtsls.setup({
  settings = { typescript = { tsserver = { maxTsServerMemory = 13312 }}}
})
```

**Comment by varunrau** [comment]
Thank you!!! That worked perfectly for me :D
